core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java

```diff
@@ -0,0 +1,392 @@
+            }
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                onError(new IllegalArgumentException(
+                        "n: " + n + " (expected: > 0, see Reactive Streams specification rule 3.9)"));
+                upstream.cancel();
+                return;
+            }
+
+            if (canceled) {
+                return;
+            }
+
+            if (executor.inEventLoop()) {
+                handleRequest(n);
+            } else {
+                executor.execute(() -> handleRequest(n));
+            }
+        }
+
+        private void handleRequest(long n) {
+            requestedByDownstream = LongMath.saturatedAdd(requestedByDownstream, n);
+            upstream.request(maxConcurrency - sourceSubscriptions.size());
+
+            flush();
+            requestAllAvailable();
+        }
+
+        @Override
+        public void cancel() {
+            upstream.cancel();
+            cancelSourceSubscriptions();
+        }
+
+        private void cancelSourceSubscriptions() {
+            sourceSubscriptions.forEach(FlatMapSubscriber::cancel);
+        }
+
+        void subscribeChild(FlatMapSubscriber<T, U> child) {
+            if (executor.inEventLoop()) {
+                handleSubscribeChild(child);
+            } else {
+                executor.execute(() -> handleSubscribeChild(child));
+            }
+        }
+
+        private void handleSubscribeChild(FlatMapSubscriber<T, U> child) {
+            pendingSubscriptions--;
+            sourceSubscriptions.add(child);
+
+            requestAllAvailable();
+        }
+
+        private long getAvailableBufferSpace() {
+            if (requestedByDownstream == Long.MAX_VALUE) {
+                return Long.MAX_VALUE;
+            }
+
+            final Optional<Long> requested = sourceSubscriptions.stream().map(FlatMapSubscriber::getRequested)
+                                                                .reduce(LongMath::saturatedAdd);
+
+            return maxConcurrency - requested.orElse(0L) - buffer.size();
+        }
+
+        private void requestAllAvailable() {
+            if (sourceSubscriptions.isEmpty()) {
+                return;
+            }
+
+            final long available = getAvailableBufferSpace();
+
+            if (available == Long.MAX_VALUE) {
+                sourceSubscriptions.forEach(sub -> sub.request(Long.MAX_VALUE));
+            }
+
+            final List<FlatMapSubscriber<T, U>> toRequest = sourceSubscriptions.stream()
+                                                                               .filter(sub -> sub.getRequested()
+                                                                                              == 0)
+                                                                               .limit(available)
+                                                                               .collect(Collectors.toList());
```

Comment:
nit:
```suggestion
                                                                               .collect(toImmutableList());
```
Yeah, creating a new list also works. 😄 
So, Could you remove `if (executor.inEventLoop()) {..} else {...}` from this class where possible? 😄

Reply:
Is that because the `FlatMapSubscriber` will run with the same executor, so we know that `completeChild` etc will always run in the same executor?

Reply:
Yes exactly.  😄 
All `onX` methods in `FlatMapAggregatingSubscriber` and `FlatMapSubscriber` are now called by the `executor` so we don't have to check it. 😄